proposals/testing/NNNN-evaluate-condition.md

```diff
@@ -0,0 +1,74 @@
+# Public API to evaluate ConditionTrait
+
+* Proposal: [SWT-NNNN](NNNN-evaluate-condition.md)
+* Authors: [David Catmull](https://github.com/Uncommon)
+* Status: **Awaiting review**
+* Implementation: [swiftlang/swift-testing#909](https://github.com/swiftlang/swift-testing/pull/909)
+* Review: ([pitch](https://forums.swift.org/t/pitch-introduce-conditiontrait-evaluate/77242))
+
+## Introduction
+
+This adds an `evaluate()` method to `ConditionTrait` to evaluate the condition
+without requiring a `Test` instance.
+
+## Motivation
+
+Currently, the only way a `ConditionTrait` is evaluated is inside the
+`prepare(for:)` method. This makes it difficult for third-party libraries to
+utilize these traits because evaluating a condition would require creating a
+dummy `Test` to pass to that method.
+
+## Proposed solution
+
+The proposal is to add a `ConditionTrait.evaluate()` method which returns the
+result of the evaluation. The existing `prepare(for:)` method is updated to call
+`evaluate()` so that the logic is not duplicated.
+
+## Detailed design
+
+The `evaluate()` method is as follows, containing essentially the same logic
+as was in `prepare(for:)`:
+
+```swift
+public func evaluate() async throws -> EvaluationResult {
+  switch kind {
+  case let .conditional(condition):
+    try await condition()
+  case let .unconditional(unconditionalValue):
+    (unconditionalValue, nil)
+  }
+}
+```
+
+`EvaluationResult` is a `typealias` for the tuple already used as the result
+of the callback in `Kind.conditional`:
+
+```swift
+public typealias EvaluationResult = (wasMet: Bool, comment: Comment?)
```

Comment:
I have a vague recollection of us discussing whether this `Comment?` was necessary in the first place but I am having trouble finding it. Did I imagine it?

Reply:
I believe that would be this message:
https://forums.swift.org/t/pitch-introduce-conditiontrait-evaluate/77242/8

Since there wasn't a conclusion on that I left it in.

Reply:
While reviewing the code I found that we can eliminate this "override comment", so I posted https://github.com/swiftlang/swift-testing/pull/1036

I think we can and should simplify this proposal by eliminating the need for the `EvaluationResult` tuple and have the `evaluate()` method just return `Bool`.

Reply:
Should I take that as authoritative, or does anyone else need to weigh in?

Reply:
Stuart and I discussed this at the Testing Workgroup meeting yesterday. While we see value in returning a customized comment, it's not actually functionality that exists today. The code inside Swift Testing to customize a comment is, in effect, dead and there's no way to actually return a custom comment here, today.

If that's something that you want to pursue, we can still do so, but we should probably do it as a separate proposal.

I hope that makes sense?